### PR TITLE
ONNX Add new pattern for densenet

### PR DESCRIPTION
**Pull request by kshpv**
### Changes

Introduce new patterns - **BatchNorm + Arithmetic Operation + Arithmetic Operation + Activation** 
**LINEAR + BatchNorm + Arithmetic Operation + Arithmetic Operation + Activation** 

image

### Reason for changes

To enable correct quantization scheme for densenet121 without ignored_scopes.

### Related tickets

No

### Tests

Add graph test for densenet121 from ONNX Model ZOO

### CODE REVIEW

File: nncf/onnx/hardware/fused_patterns.py [hunk 1]
```diff
@@ -52,8 +52,12 @@ def _get_onnx_hw_fused_patterns() -> HWFusedPatterns:
                                          activations + batch_norms | \
                                          batch_norms | activations
 
+    batch_norm_arithmetic_arithmetic_act = batch_norms + arithmetic_ops + arithmetic_ops + activations
```
Comment (KodiaqQ) [inline nncf/onnx/hardware/fused_patterns.py:55]: Let's add new patterns in the proper and clean way: with the particular method, `add_node` and `add_edge` functionality, for each node, with the needed type (not **ARITHMETIC_OPS). It's more convenient to support, readable, and easier to imagine/describe graphically. 
For now, you've added the A*B*C*D = E numbers of patterns, while initially describing only one that is really needed.
Additionally, the original pattern looks like a `batch norm layer + scale shift pattern` and not as `BN_ARITHMETHIC_ARITHMETHIC_ACT`, `LINEAR + BN_ARITHMETHIC_ARITHMETHIC_ACT_`, ..., etc.

File: tests/onnx/quantization/test_classification_models_graph.py [hunk 2]
```diff
@@ -17,29 +17,31 @@
...
 # pylint: disable=no-member
-
+from tests.onnx.conftest import ONNX_MODEL_DIR
 from tests.onnx.quantization.common import ModelToTest
 from tests.onnx.quantization.common import min_max_quantize_model
 from tests.onnx.quantization.common import compare_nncf_graph
 from tests.onnx.quantization.common import mock_collect_statistics
+from tests.onnx.weightless_model import load_model_topology_with_zeros_weights
+
+TORCHVISION_TEST_DATA = [(ModelToTest('resnet18', [1, 3, 224, 224]), models.resnet18(pretrained=True)),
```
Comment (KodiaqQ) [inline tests/onnx/quantization/test_classification_models_graph.py:27]: Why did you change the test and the reference?

File: tests/onnx/quantization/test_classification_models_graph.py [hunk 3]
```diff
@@ -48,3 +50,17 @@ def test_min_max_quantization_graph(tmp_path, mocker, model_to_test, model):
...
     quantized_model = min_max_quantize_model(model_to_test.input_shape, original_model)
     compare_nncf_graph(quantized_model, model_to_test.path_ref_graph)
+
+
+ONNX_TEST_DATA = [ModelToTest('densenet-12', [1, 3, 224, 224])]
+
+
+@pytest.mark.parametrize(('model_to_test'), ONNX_TEST_DATA,
+                         ids=[model_to_test.model_name for model_to_test in ONNX_TEST_DATA])
+def test_min_max_quantization_graph_onnx_model(tmp_path, mocker, model_to_test):
```
Comment (KodiaqQ) [inline tests/onnx/quantization/test_classification_models_graph.py:60]: Why do we need this test now?

File: tests/onnx/quantization/test_classification_models_graph.py (hunk 2 above)
Comment (kshpv) [reply tests/onnx/quantization/test_classification_models_graph.py:27]: I add additional test for densenet model from onnx model zoo, which has the pattern that I added. Because of that I divided the test onto two: models from torchvision and onnx models

File: tests/onnx/quantization/test_classification_models_graph.py (hunk 3 above)
Comment (kshpv) [reply tests/onnx/quantization/test_classification_models_graph.py:60]: Because densenet model has the pattern that I added, I want to test it by this test

File: nncf/onnx/hardware/fused_patterns.py (hunk 1 above)
Comment (kshpv) [reply nncf/onnx/hardware/fused_patterns.py:55]: I agree with you that `add_node` and `add_edge` is more explicit way to create patterns.
At the same time there are two patterns that should be added: 
BN + MUL + ADD + ACTIVATION and LINEAR + MUL + ADD +ACTIVATION, as both of them are faced in the model

image

Comment (KodiaqQ) [reply nncf/onnx/hardware/fused_patterns.py:55]: You can add only one pattern like:
```suggestion
pattern = GraphPattern()
batchnorm_node = pattern.add_node(GraphPattern.LABEL_ATTR: 'BATCHNORM', GraphPattern.METATYPE_ATTR: om.ONNXBatchNormMetatype)
mul_node = pattern.add_node(GraphPattern.LABEL_ATTR: 'MULTIPLY', GraphPattern.METATYPE_ATTR: om.ONNXMulLayerMetatype)
add_node = pattern.add_node(GraphPattern.LABEL_ATTR: 'ADD', GraphPattern.METATYPE_ATTR: om.ONNXAddLayerMetatype)
pattern.add_edge(batchnorm_node, mul_node)
pattern.add_edge(mul_node, add_node)
```
instead of the overburdened and unreadable list of `BN + MUL + ADD + ACTIVATION` and `LINEAR + MUL + ADD +ACTIVATION`

Comment (kshpv) [reply nncf/onnx/hardware/fused_patterns.py:55]: Again - I need two patterns: `linear_ops + batch_norm + scale_shift + activations` and `batch_norm + scale_shift + activations`
I propose to use the already created `scale_shift` pattern and using `+` operator to connect them.
`+` - is easy understandable operation

Comment (KodiaqQ) [reply nncf/onnx/hardware/fused_patterns.py]: If you need two patterns, then add two patterns with the two different methods with different labels, instead of using the `+` operators.

Comment (KodiaqQ) [reply nncf/onnx/hardware/fused_patterns.py]: As you wish. I'm not insisting.

Comment (kshpv) [reply nncf/onnx/hardware/fused_patterns.py]: I hope other reviewers will add their minds here and we can make a finall decision whether should we add patterns only with `add_node` and `add_edge` methods.

**Comment by kshpv** [comment]
run onnx pre-commit tests

**Review by vshampor — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]
LGTM

**Review by daniil-lyakhov — APPROVED** [review]

**Review by l-bat — APPROVED** [review]

File: nncf/onnx/hardware/fused_patterns.py (hunk 1 above)
Comment (kshpv) [reply nncf/onnx/hardware/fused_patterns.py]: I've updated the patterns without + and | operators.

File: nncf/common/graph/patterns.py [hunk 4]
```diff
@@ -107,9 +107,9 @@ def __add__(self, other: 'GraphPattern') -> 'GraphPattern':
                 # add merge all possible connections
                 # A: (a) (b)
                 # B: (c) (d)
-                #              (a)  (a_copy)  (b)    (b_copy)
+                #              (a_copy)  (a_copy)  (b_copy)    (b_copy)
                 # A + B ---->   |       |      |        |
-                #              (c)     (d)  (c_copy) (d_copy)
+                #              (c_copy)     (d_copy)  (c_copy) (d_copy)
```
Comment (andrey-churkin) [inline nncf/common/graph/patterns.py:112]: Is it the correct change?

Comment (kshpv) [reply nncf/common/graph/patterns.py:112]: The previous comment was incorrect, I fixed it.
In `__add__` magic method we create a new graph and copy every weakly component for both initial graphs (self, other), connect them and put into the final graph.
In this logic all weakly connected components are copies.

Comment (andrey-churkin) [reply nncf/common/graph/patterns.py:112]: Seems like we need to align `|` as well.

Comment (kshpv) [reply nncf/common/graph/patterns.py]: Done

**Review by vshampor — APPROVED** [review]

File: nncf/common/graph/patterns.py [hunk 5]
```diff
@@ -194,9 +194,16 @@ def join_patterns(self, other: 'GraphPattern',
         """
         Adds 'other' pattern to 'self' pattern and connect nodes from self to other specified by 'edges'.
 
-        If edges is None, adds an edge between
-        last node of self's graph and first node of other's graph,
-        which are found by nx.lexicographical_topological_sort().
+        If edges is None, connects all weekly connected components of self and other by adding edges between
+        last nodes of every weekly component of self and first nodes of every weekly component other.
```
Comment (l-bat) [inline nncf/common/graph/patterns.py:198]:
```suggestion
        If edges is None, connects all weakly connected components of self and other by adding edges between
        last nodes of every weakly component of self and first nodes of every weakly component other.
```

**Review by l-bat — APPROVED** [review]

**Review by KodiaqQ — APPROVED** [review]

**Review by andrey-churkin — APPROVED** [review]

**Comment by kshpv** [comment]
run pytorch pre-commit tests

**Comment by kshpv** [comment]
run pytorch pre-commit tests

**Review by alexsu52 — APPROVED** [review]
LGTM

**Review by AlexKoff88 — APPROVED** [review]
